corehq/motech/repeaters/tasks.py

```diff
@@ -72,65 +64,103 @@ def delete_old_request_logs():
-        check_repeaters_in_partition.delay(current_partition)
-
-
-@task(queue=settings.CELERY_PERIODIC_QUEUE)
-def check_repeaters_in_partition(partition):
-    """
-    The CHECK_REPEATERS_PARTITION_COUNT constant dictates the total number of partitions
-    :param partition: index of partition to check
-    """
     start = datetime.utcnow()
     twentythree_hours_sec = 23 * 60 * 60
     twentythree_hours_later = start + timedelta(hours=23)
 
-    # Long timeout to allow all waiting repeat records to be iterated
-    lock_key = f"{CHECK_REPEATERS_KEY}_{partition}_in_{CHECK_REPEATERS_PARTITION_COUNT}"
+    # Long timeout to allow all waiting repeaters to be iterated
+    # TODO: Check how long it takes to iterate all repeaters on prod
     check_repeater_lock = get_redis_lock(
-        lock_key,
+        CHECK_REPEATERS_KEY,
         timeout=twentythree_hours_sec,
-        name=lock_key,
+        name=CHECK_REPEATERS_KEY,
     )
     if not check_repeater_lock.acquire(blocking=False):
-        metrics_counter("commcare.repeaters.check.locked_out", tags={'partition': partition})
+        metrics_counter("commcare.repeaters.check.locked_out")
         return
 
     try:
         with metrics_histogram_timer(
             "commcare.repeaters.check.processing",
             timing_buckets=_check_repeaters_buckets,
         ):
-            for record in RepeatRecord.objects.iter_partition(
-                    start, partition, CHECK_REPEATERS_PARTITION_COUNT):
-                if not _soft_assert(
-                    datetime.utcnow() < twentythree_hours_later,
-                    "I've been iterating repeat records for 23 hours. I quit!"
-                ):
-                    break
-
+            for repeater in iter_ready_repeaters():
                 metrics_counter("commcare.repeaters.check.attempt_forward")
-                record.attempt_forward_now(is_retry=True)
-            else:
-                iterating_time = datetime.utcnow() - start
-                _soft_assert(
-                    iterating_time < timedelta(hours=6),
-                    f"It took {iterating_time} to iterate repeat records."
-                )
+                process_repeater.delay(repeater.domain, repeater.repeater_id)
+
+                if datetime.utcnow() > twentythree_hours_later:
+                    # Break after process_repeater() to avoid a stale lock
+                    break
     finally:
         check_repeater_lock.release()
 
 
+def iter_ready_repeaters():
+    """
+    Cycles through repeaters (repeatedly ;) ) until there are no more
+    repeat records ready to be sent.
+    """
+    while True:
+        yielded = False
+        for repeater in Repeater.objects.all_ready():
+            if not repeater.domain_can_forward:
+                continue
+
+            if rate_limit_repeater(repeater.domain):
+                repeater.rate_limit()
+                continue
+
+            yielded = True
+            yield repeater
```

Comment:
It looks like this may yield the same repeater multiple times in a single iteration (because of `while True`)... which is making me think that it's no problem to "have two `process_repeater` tasks running concurrently for the same repeater" (as I asked above), or maybe that was an an oversight here?

Reply:
(See other [thread on concurrency](https://github.com/dimagi/commcare-hq/pull/34946#discussion_r1707919900).)